File: python/vista_api_client/api_resources/usersets.py

from vista_api_client.api_resources.api_resource import ApiResource, HttpMethods
from vista_api_client.api_resources.grants import Grants


class Usersets(ApiResource):
    def __init__(self, secret, branch, hostname):
        super().__init__(secret, branch, hostname)
        self.grants = Grants(self.secret, self.branch, self.hostname)

    def create(self, userset_id, org_id, parent_usersets=None):
        parents = parent_usersets or []
        return self.dispatch('/v1/usersets', HttpMethods.POST, {
            'id': userset_id,
            'org_id': org_id,
            'parent_usersets': parents,
            'branch': self.branch,
        })

    def inherit(self, child_userset_id, parent_userset_id):
        return self.dispatch('/v1/usersets/inherit', HttpMethods.POST, {
            'child_userset_id': child_userset_id,
            'parent_userset_id': parent_userset_id,
            'branch': self.branch,
        })

    def grant_action(self, user_id, action, resource_id, resource_type, attribute=''):
        return self.grants.grant(user_id, 'USERSET', action, 'ACTION',
                                 resource_id, resource_type, attribute)

    def revoke_action(self, user_id, action, resource_id, resource_type, attribute=''):
        return self.grants.revoke(user_id, 'USERSET', action, 'ACTION',
                                  resource_id, resource_type, attribute)

    def grant_role(self, user_id, role_id, resource_id, resource_type):
        return self.grants.grant(user_id, 'USERSET', role_id, 'ROLE',
                                 resource_id, resource_type, '')

    def revoke_role(self, user_id, role_id, resource_id, resource_type):
        return self.grants.revoke(user_id, 'USERSET', role_id, 'ROLE',
                                  resource_id, resource_type, '')

File: js/vista-api-client/src/resources/Usersets.ts

import { AxiosInstance } from 'axios';
import { HttpMethods, ApiResource } from '../apiResource.js';
import Grants from './Grants';

class Usersets extends ApiResource {
    grants: Grants;

    constructor(axiosClient: AxiosInstance, branch: string, hostname: string) {
        super(axiosClient, branch, hostname);
        this.grants = new Grants(axiosClient, branch, hostname);
    }

    create = async (usersetId: string, orgId: string, parentUsersets=[]) => {
        return this.dispatch('/v1/usersets', HttpMethods.POST, {
            id: usersetId,
            org_id: orgId,
            parent_usersets: parentUsersets,
            branch: this.branch,
        });
    }

    inherit = async (childUsersetId: string, parentUsersetId: string) => {
        return this.dispatch(`/v1/usersets/inherit`, HttpMethods.POST, {
            child_userset_id: childUsersetId,
            parent_userset_id: parentUsersetId,
            branch: this.branch,
        });
    }

    grantAction = async (usersetId: string, action: string, resourceId: string, resourceType: string, attribute = '') => {
        return this.grants.grant(usersetId, 'USERSET', action, 'ACTION', resourceId, resourceType, attribute);
    }

    revokeAction = async (usersetId: string, action: string, resourceId: string, resourceType: string, attribute = '') => {
        return this.grants.revoke(usersetId, 'USERSET', action, 'ACTION', resourceId, resourceType, attribute);
    }

    grantRole = async (usersetId: string, role_id: string, resourceId: string, resourceType: string) => {
        return this.grants.grant(usersetId, 'USERSET', role_id, 'ROLE', resourceId, resourceType, '');
    }

    revokeRole = async (usersetId: string, role_id: string, resourceId: string, resourceType: string) => {
        return this.grants.revoke(usersetId, 'USERSET', role_id, 'ROLE', resourceId, resourceType, '');
    }
}

export default Usersets;
